internal/overlord/logstate/loki.go

```diff
@@ -0,0 +1,250 @@
+) (*lokiForwarder, error) {
+
+	labels := map[string]string{
+		"pebble_service": serviceName,
+	}
+
+	return &lokiForwarder{
+		serviceName: serviceName,
+		targetName:  targetName,
+		target:      target,
+		iterator:    iterator,
+		parser:      servicelog.NewParser(iterator, lokiParserSize),
+		buffer:      newLokiBuffer(labels, 10, 1*time.Second),
+		client:      newLokiClient(target.Location),
+	}, nil
+}
+
+func (f *lokiForwarder) Forward() {
+	for {
+		// Check if Stop has been called
+		select {
+		case <-f.stopCh:
+			f.flushBuffer()
+			return
+		default:
+			// continue
+		}
+
+		if f.buffer.ReadyToFlush() {
+			f.flushBuffer()
+		}
+
+		if f.parser.Next() {
+			f.buffer.WriteEntry(f.parser.Entry())
+		} else {
+			if err := f.parser.Err(); err != nil {
+				logger.Noticef(
+					"Forwarding logs to target %q: cannot parse log for service %q: %v",
+					f.targetName, f.serviceName, err)
+			}
+		}
+	}
+}
+
+// Alternative implementation suggested by Harry, however it doesn't work.
+// f.iterator.Next(f.stopCh) hangs and never returns a value.
+//func (f *lokiForwarder) Forward() {
+//	for f.iterator.Next(f.stopCh) {
+//		for f.parser.Next() {
+//			f.buffer.WriteEntry(f.parser.Entry())
+//		}
+//		if err := f.parser.Err(); err != nil {
+//			logger.Noticef(
+//				"Forwarding logs to target %q: cannot parse log for service %q: %v",
+//				f.targetName, f.serviceName, err)
+//			// TODO: create a new parser?
+//		}
+//		if !f.buffer.ReadyToFlush() {
+//			f.flushBuffer()
+//		}
+//	}
+//
+//	// Forwarder stopped or service exited - flush remaining logs
+//	if len(f.buffer.messages) > 0 {
+//		f.flushBuffer()
+//	}
+//}
+
+func (f *lokiForwarder) Stop() {
+	f.stopCh <- struct{}{}
+}
+
+func (f *lokiForwarder) flushBuffer() {
+	err := f.client.Send(f.buffer.Request())
+	if err != nil {
+		logger.Noticef(
+			"Forwarding logs to target %q: cannot send logs for service %q: %v",
+			f.targetName, f.serviceName, err)
+		// TODO: retry?
+	}
+
+	f.buffer.Reset()
+}
+
+type lokiBuffer struct {
+	labels map[string]string
+
+	buffer  *bytes.Buffer
+	encoder *json.Encoder
+
+	messages    int
+	maxMessages int
+	deadline    *time.Time
+	delay       time.Duration
+}
+
+func newLokiBuffer(labels map[string]string, maxMessages int, delay time.Duration) *lokiBuffer {
+	buf := bytes.NewBuffer(make([]byte, 0, 4096))
+
+	b := &lokiBuffer{
+		labels:      labels,
+		buffer:      buf,
+		encoder:     json.NewEncoder(buf),
+		maxMessages: maxMessages,
+		delay:       delay,
+	}
+	b.Reset()
+	return b
+}
+
+func (b *lokiBuffer) Reset() {
+	b.messages = 0
+	b.deadline = nil
+	b.buffer.Reset()
+
+	// Write start of request
+	b.buffer.WriteString(`{"streams":[{"stream":`)
+	b.encoder.Encode(b.labels)
+	b.buffer.WriteString(`,"values":[`)
+}
+
+func (b *lokiBuffer) WriteEntry(entry servicelog.Entry) {
+	// Write in format
+	//    ["<unix epoch in nanoseconds>","<log line>"]
+	b.buffer.WriteString("[")
```

Comment:
All these lines could be trimmed down to just an fmt style command?

Reply:
The fmt commands are much less efficient - see Ben's review comments